staging/src/k8s.io/apimachinery/pkg/util/httpstream/spdy/roundtripper.go

```diff
@@ -273,20 +287,50 @@ func (s *SpdyRoundTripper) tlsConn(ctx context.Context, rwc net.Conn, targetHost
 // dialWithoutProxy dials the host specified by url, using TLS if appropriate.
 func (s *SpdyRoundTripper) dialWithoutProxy(ctx context.Context, url *url.URL) (net.Conn, error) {
 	dialAddr := netutil.CanonicalAddr(url)
-	dialer := s.Dialer
-	if dialer == nil {
-		dialer = &net.Dialer{}
+	dial := s.DialContext
+	// prefer DialContext
+	if dial == nil {
+		// but fallback to the provided Dialer and then a generic net.Dialer
+		if s.Dialer != nil {
+			dial = s.Dialer.DialContext
+		} else {
+			nd := &net.Dialer{}
+			dial = nd.DialContext
+		}
 	}
 
+	conn, err := dial(ctx, "tcp", dialAddr)
+	if err != nil {
+		return nil, err
+	}
 	if url.Scheme == "http" {
-		return dialer.DialContext(ctx, "tcp", dialAddr)
+		return conn, nil
+	}
+
+	// This code is identical to tls.Dial but replicated here so we can
+	// use our own DialContext function.
+	tlsConfig := s.tlsConfig
+	if tlsConfig == nil {
+		tlsConfig = &tls.Config{}
+	}
+
+	// If no ServerName is set, infer the ServerName
+	// from the hostname we're connecting to.
+	if tlsConfig.ServerName == "" {
+		tlsConfig = tlsConfig.Clone()
+		colonPos := strings.LastIndex(dialAddr, ":")
+		if colonPos == -1 {
+			colonPos = len(dialAddr)
+		}
+		tlsConfig.ServerName = dialAddr[:colonPos]
 	}
```

Comment:
it's *really* unfortunate we have to duplicate this and manage handshaking ourselves... there's no way to use stdlib tls setup with a custom dialer?

Reply:
Agreed, but I don't see a way around it given the current state of things:

```go
package tls

func Dial(network, addr string, config *Config) (*Conn, error) {
	return DialWithDialer(new(net.Dialer), network, addr, config)
}
func DialWithDialer(dialer *net.Dialer, network, addr string, config *Config) (*Conn, error) {
	return dial(context.Background(), dialer, network, addr, config)
}
func dial(ctx context.Context, netDialer *net.Dialer, network, addr string, config *Config) (*Conn, error) {
...
```

There's an issue https://github.com/golang/go/issues/9360 that suggests using a standard dialer interface throughout the Go stdlib when only dial methods are used.  I can't find any specific language that prohibits replacing a concrete type with an interface with respect to Go1 compatibility, but I suspect it's considered a breaking change as the issue is tagged v2.